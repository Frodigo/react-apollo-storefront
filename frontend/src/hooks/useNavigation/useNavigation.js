import { useMemo } from 'react';
import { useQuery } from '@apollo/client';
import { GET_NAVIGATION } from '../../queries/category'

/**
 * The useNavigation hook provides data and business logic for the Navigation component
 *
 *
 * @return {
 *  navigationData {array} - array of categories to display on a storefront
 * }
 */
export const useNavigation = () => {
    /**
     * Get category data from API
     */
    const { data } = useQuery(GET_NAVIGATION, {
        fetchPolicy: 'cache-and-network',
        nextFetchPolicy: 'cache-first'
    });

    /**
     * Check if category should be visible on the storefront.
     *
     * @param {NavigationItem} category
     * @returns {boolean}
     */
    const shouldRenderNavigationItem = category => {
        return !!category.include_in_menu;
    };


    /**
     * Preparing navigation data
     * Sorts by category.position
     * Filtering by category.include_in_menu flag
     * Adds categoryUrl field
     *
     * @returns {array}
     */
    const navigationData = useMemo(() => {
        const rootCategory = data && data.categoryList ? data.categoryList[0] : null;

        if (rootCategory) {
            return rootCategory.children
                .filter(category => shouldRenderNavigationItem(category))
                .sort((a, b) => (a.position > b.position ? 1 : -1))
                .map(category => {
                    const categoryUrl = `/category/${category.url_path}${category.url_suffix}`

                    return {
                        ...category,
                        categoryUrl
                    }
                })
        }

        return [];
    }, [data]);

    return {
        navigationData
    }
}

export default useNavigation;
