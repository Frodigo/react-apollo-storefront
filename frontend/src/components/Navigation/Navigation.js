import React from 'react';
import { Navbar, Nav, Container } from 'react-bootstrap';

import useNavigation from '../../hooks/useNavigation';

const Navigation = () => {
    const { navigationData } = useNavigation();
    const navigationItems = navigationData.map(navItem =>
        <Nav.Item key={navItem.id}>
            <Nav.Link href={navItem.categoryUrl}>{navItem.name}</Nav.Link>
        </Nav.Item>
    );

    const shouldDisplayNavigation = navigationItems && navigationItems.length ? <Navbar bg="light" expand="lg">
        <Container>
        <Navbar.Toggle aria-controls="primary-nav" />
            <Navbar.Collapse id="primary-nav">
                <Nav className="mr-auto" activeKey="/">
                    {navigationItems}
                </Nav>
            </Navbar.Collapse>
        </Container>
    </Navbar> : null;

    return shouldDisplayNavigation;
};

export default Navigation;
