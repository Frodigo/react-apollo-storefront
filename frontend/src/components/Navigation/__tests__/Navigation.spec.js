import React from 'react';
import { render } from '@testing-library/react';

import Navigation from '../Navigation';
import useNavigation from '../../../hooks/useNavigation';

jest.mock('../../../hooks/useNavigation');

test('Navigation renders correctly', () => {
    useNavigation.mockReturnValueOnce({
        navigationData:  [
            {
                "__typename": "CategoryTree",
                "id": 38,
                "include_in_menu": 1,
                "name": "What's New",
                "position": 1,
                "url_path": "what-is-new",
                "url_suffix": ".html",
                "categoryUrl": "/category/what-is-new.html"
            },
            {
                "__typename": "CategoryTree",
                "id": 20,
                "include_in_menu": 1,
                "name": "Women",
                "position": 2,
                "url_path": "women",
                "url_suffix": ".html",
                "categoryUrl": "/category/women.html"
            },
            {
                "__typename": "CategoryTree",
                "id": 11,
                "include_in_menu": 1,
                "name": "Men",
                "position": 3,
                "url_path": "men",
                "url_suffix": ".html",
                "categoryUrl": "/category/men.html"
            },
            {
                "__typename": "CategoryTree",
                "id": 3,
                "include_in_menu": 1,
                "name": "Gear",
                "position": 4,
                "url_path": "gear",
                "url_suffix": ".html",
                "categoryUrl": "/category/gear.html"
            },
            {
                "__typename": "CategoryTree",
                "id": 9,
                "include_in_menu": 1,
                "name": "Training",
                "position": 5,
                "url_path": "training",
                "url_suffix": ".html",
                "categoryUrl": "/category/training.html"
            },
            {
                "__typename": "CategoryTree",
                "id": 37,
                "include_in_menu": 1,
                "name": "Sale",
                "position": 6,
                "url_path": "sale",
                "url_suffix": ".html",
                "categoryUrl": "/category/sale.html"
            }
        ]
    })
    const { container } = render(<Navigation />);

    expect(container).toMatchSnapshot();
});

test('Navigation not renders if there is no items', () => {
    useNavigation.mockReturnValueOnce({
        navigationData: []
    })
    const { container } = render(<Navigation />);

    expect(container.firstChild).toBeNull();
});
