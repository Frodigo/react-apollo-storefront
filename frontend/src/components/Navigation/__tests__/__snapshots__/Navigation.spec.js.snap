// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation renders correctly 1`] = `
<div>
  <nav
    class="navbar navbar-expand-lg navbar-light bg-light"
  >
    <div
      class="container"
    >
      <a
        class="d-lg-none"
        href="/"
      >
        React Apollo
      </a>
      <button
        aria-controls="primary-nav"
        aria-label="Toggle navigation"
        class="navbar-toggler collapsed"
        type="button"
      >
        <span
          class="navbar-toggler-icon"
        />
      </button>
      <div
        class="navbar-collapse collapse"
        id="primary-nav"
      >
        <div
          class="mr-auto navbar-nav"
        >
          <div
            class="nav-item"
          >
            <a
              class="nav-link"
              data-rb-event-key="/category/what-is-new.html"
              href="/category/what-is-new.html"
            >
              What's New
            </a>
          </div>
          <div
            class="nav-item"
          >
            <a
              class="nav-link"
              data-rb-event-key="/category/women.html"
              href="/category/women.html"
            >
              Women
            </a>
          </div>
          <div
            class="nav-item"
          >
            <a
              class="nav-link"
              data-rb-event-key="/category/men.html"
              href="/category/men.html"
            >
              Men
            </a>
          </div>
          <div
            class="nav-item"
          >
            <a
              class="nav-link"
              data-rb-event-key="/category/gear.html"
              href="/category/gear.html"
            >
              Gear
            </a>
          </div>
          <div
            class="nav-item"
          >
            <a
              class="nav-link"
              data-rb-event-key="/category/training.html"
              href="/category/training.html"
            >
              Training
            </a>
          </div>
          <div
            class="nav-item"
          >
            <a
              class="nav-link"
              data-rb-event-key="/category/sale.html"
              href="/category/sale.html"
            >
              Sale
            </a>
          </div>
        </div>
      </div>
    </div>
  </nav>
</div>
`;
